### ISSUE TITLE
Log function causes floating point underflow/ gradient overflow

### ISSUE THREAD

**Issue by zhangsh950618** [NVIDIA/apex#379]
we use softmax->log->nll_loss instead of nn.CrossEntropyLoss() to calculate the cross entropy loss. But when training with fp16_optimizer, we found sudden gradient overflows and skip all steps after training several batches. We debug for several hours and read the source code, we  found the log function could cause the floating point underflow and gradient overflow as the apex could caluculate the norm of the gradient to determine weather the gradient overflow.

**Comment by ptrblck** [comment]
Hi @zhangsh950618,

are you applying `softmax` and `log` separately when you encounter the gradient overflows?
If so, do you only see the gradient overflows using `amp` with `opt_level!=O0` or also in a vanilla FP32 training?

It is generally recommended to use `F.log_softmax` instead of `torch.log(F.softmax())`, as the latter is slower and numerically unstable.

**Comment by mcarilli** [comment]
Don't use FP16_Optimizer, use [the current API](https://nvidia.github.io/apex/amp.html#opt-levels-and-properties).  opt_level="O1" is recommended.

**Comment by huijiefeng** [comment]
I generally encountered the same problem. For some algorithmic purposes I need to apply `softmax` and `log` separately. This works well in standard full precision training, but will cause nan outputs due to gradient overflows after using `amp` with `opt_level = O2`. I'm thinking if manual scaling during the training process would resolve this issue.
